osu.Game.Rulesets.Osu/Statistics/AccuracyHeatmap.cs

```diff
@@ -250,48 +257,73 @@ protected void AddPoint(Vector2 start, Vector2 end, Vector2 hitPoint, float radi
             var rotatedCoordinate = -1 * new Vector2((float)Math.Cos(rotatedAngle), (float)Math.Sin(rotatedAngle));
 
             Vector2 localCentre = new Vector2(points_per_dimension - 1) / 2;
-            float localRadius = localCentre.X * inner_portion * normalisedDistance; // The radius inside the inner portion which of the heatmap which the closest point lies.
+            float localRadius = localCentre.X * inner_portion * normalisedDistance;
             Vector2 localPoint = localCentre + localRadius * rotatedCoordinate;
 
             // Find the most relevant hit point.
-            int r = Math.Clamp((int)Math.Round(localPoint.Y), 0, points_per_dimension - 1);
-            int c = Math.Clamp((int)Math.Round(localPoint.X), 0, points_per_dimension - 1);
+            int r = (int)Math.Round(localPoint.Y);
+            int c = (int)Math.Round(localPoint.X);
+
+            if (r < 0 || r >= points_per_dimension || c < 0 || c >= points_per_dimension)
+                return;
 
-            PeakValue = Math.Max(PeakValue, ((HitPoint)pointGrid.Content[r][c]).Increment());
+            PeakValue = Math.Max(PeakValue, ((GridPoint)pointGrid.Content[r][c]).Increment());
```

Comment:
So on another look I'm not entirely sure this does what it's supposed to be doing?

Why? Because you can _be inside the hitcircle and still miss_ if you hit the circle too early.

On this branch, the following can happen:

https://github.com/ppy/osu/assets/20418176/0ff6ced3-67fa-4e2e-b31b-c32be1e3d6d4

I missed all three circles due to hitting too early but did not fail because HP didn't drain enough. The accuracy heatmap is expectedly empty. On the heatmap, two hits are registered, even though _I didn't actually hit anything_.

I'm not sure what to do with this or if this is undesirable even, but it does seem counterintuitive at least. I would probably just not count such hits.

Reply:
Personally I think that's fine.